From ebcebe2ead4911f9e6d96f867b8736a833b5d1cb Mon Sep 17 00:00:00 2001
From: benno <benno@openbsd.org>
Date: Fri, 22 May 2020 07:50:07 +0000
Subject: [PATCH] Fix the exit code when eval()uating a || compound list, it
 would terminate the shell when running under -e. See also
 https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=269067 and Bug reported
 including fix by Leah Neukirchen, Thanks! ok millert@

---
 c_sh.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/c_sh.c b/c_sh.c
index b051022..98f5e5e 100644
--- a/c_sh.c
+++ b/c_sh.c
@@ -1,4 +1,4 @@
-/*	$OpenBSD: c_sh.c,v 1.63 2018/04/09 17:53:36 tobias Exp $	*/
+/*	$OpenBSD: c_sh.c,v 1.64 2020/05/22 07:50:07 benno Exp $	*/
 
 /*
  * built-in Bourne commands
@@ -424,6 +424,8 @@ int
 c_eval(char **wp)
 {
 	struct source *s;
+	struct source *saves = source;
+	int savef;
 	int rv;
 
 	if (ksh_getopt(wp, &builtin_opt, null) == '?')
@@ -458,7 +460,11 @@ c_eval(char **wp)
 		exstat = subst_exstat;
 	}
 
+	savef = Flag(FERREXIT);
+	Flag(FERREXIT) = 0;
 	rv = shell(s, false);
+	Flag(FERREXIT) = savef;
+	source = saves;
 	afree(s, ATEMP);
 	return (rv);
 }
-- 
2.27.0

